### ISSUE TITLE
Discussing (and cleaning up) default options passed to Z3

### ISSUE THREAD

**Issue by zvonimir** [boogie-org/boogie#73]
By default, Boogie passes the following options to Z3 through queries:
```
(set-option :print-success false)
(set-info :smt-lib-version 2.0)
(set-option :AUTO_CONFIG false)
(set-option :pp.bv_literals false)
(set-option :MODEL.V2 true)
(set-option :smt.PHASE_SELECTION 0)
(set-option :smt.RESTART_STRATEGY 0)
(set-option :smt.RESTART_FACTOR |1.5|)
(set-option :smt.ARITH.RANDOM_INITIAL_VALUE true)
(set-option :smt.CASE_SPLIT 3)
(set-option :smt.DELAY_UNITS true)
(set-option :NNF.SK_HACK true)
(set-option :smt.MBQI false)
(set-option :smt.QI.EAGER_THRESHOLD 100)
(set-option :TYPE_CHECK true)
(set-option :smt.BV.REFLECT true)
```
I first used Boogie probably 10 years ago or so, and some of these options if I remember correctly were already there :). I wonder if it is time to revisit these given that Z3 changed a lot since then. Ideally, it would be good to let Z3 choose what the optimal setup should be by enabling `AUTO_CONFIG`.
Of course, such a change might break some regressions by causing time outs, but it might be a good move in the long run.
I should also add that these options are I think geared towards particular kinds of benchmarks (quantifier-heavy, etc.) and they are likely not a good setup for other usage scenarios. So I think that going with default Z3 options would be a good thing to do, while users can on top of that customize by adding options that are good for their benchmarks.
Thoughts?

**Comment by shazqadeer** [comment]
@zvonimir :  Looks like there is this old issue that has resurfaced.  Let us discuss here and do the necessary cleanup.

@bkragl : FYI.

**Comment by zvonimir** [comment]
Boogie is currently used in more scenarios than a decade ago, and various usage scenarios might require different Z3 parameters to be set. So I wonder if the code that sets the parameters should be refactored to reflect this. One idea I has in this direction was that by default no Z3 parameters get set except for the absolutely necessary ones (such as model generation), which would leave Z3 to select its default parameters based on the generated query. Then, we could have configuration files from which other parameters would be loaded as needed (Daisy would have one, Corral one, maybe CIVL one).

Finally, there is an issue of how to make sure we are not breaking anything by doing this (maintaining backwards performance and completeness "compatibility"), which I think mainly boils down to performance and completeness issues. I guess once this change is made, all interested parties should test it with their usage scenarios and report back.

**Comment by zvonimir** [comment]
Another interesting issue with Z3 parameters I identified while back is described in #53:
Recently, we observed that Boogie seems to silently drop Z3 options when they are passed to it using lower-case letters, as in `/z3opt:smt.mbqi=true`. We tracked this behavior to this interesting piece of logic that deals with Z3 options:
https://github.com/boogie-org/boogie/blob/d93c222b70ebefa01d2f25685b95b2005697c9f3/Source/Provers/SMTLib/Z3.cs#L374
It appears that the options are filtered based on whether the first character is a capital letter or not, which is very strange since Z3 options are case insensitive. For example, if Boogie is invoked with `/z3opt:SMT.mbqi=true` the option makes its way into the generated query, while if Boogie is invoked with `/z3opt:smt.mbqi=true` it is unclear what happens in that case, but the behaviors are not identical.
My hunch is that this might be legacy code that should be cleaned up, but I am not totally sure, and so help would be appreciated.

**Comment by bkragl** [comment]
As a first step I'll try to run Boogie on all the regression tests without setting any Z3 options, and produce a spreadsheet with runtime differences.

**Comment by zvonimir** [comment]
Good idea.
Btw, I also vaguely remember that Boogie used to set some Z3 parameters by adding them into the query itself, and others through command line when Z3 would get invoked. I am not sure if that's still the case. It would be good to switch to using just the query mechanism since it is more transparent.

**Comment by zvonimir** [comment]
Just brainstorming, but another idea would be to extend Boogie syntax with `set-solver-parameter` or something like that, and be able to do it directly from Boogie files. I probably like this better than the config files solution I mentioned above.

**Comment by bkragl** [comment]
Using Z3 4.8.6 and passing no options at all results in 59 tests that fail and 32 tests that diverge. Also, I found that 5 of the failing test also fail with the options that Boogie sets currently.

**Failing**
```
bitvectors/bv5.bpl
extractloops/detLoopExtract.bpl
extractloops/detLoopExtract1.bpl
extractloops/detLoopExtract2.bpl
extractloops/detLoopExtractNested.bpl
extractloops/t1.bpl
extractloops/t2.bpl
extractloops/t3.bpl
inline/test0.bpl
prover/z3mutl.bpl
stratifiedinline/bar1.bpl
stratifiedinline/bar10.bpl
stratifiedinline/bar11.bpl
stratifiedinline/bar12.bpl
stratifiedinline/bar13.bpl
stratifiedinline/bar2.bpl
stratifiedinline/bar3.bpl
stratifiedinline/bar4.bpl
stratifiedinline/bar6.bpl
stratifiedinline/bar7.bpl
stratifiedinline/bar8.bpl
stratifiedinline/bar9.bpl
symdiff/foo.bpl
test15/CaptureState.bpl
test15/IntInModel.bpl
test15/ModelTest.bpl
test15/NullInModel.bpl
test2/LambdaExt.bpl
test2/LambdaPoly.bpl
test2/NeverPattern.bpl
test2/Quantifiers.bpl
test2/Rlimitouts0.bpl
test2/SelectiveChecking.bpl
test2/UpdateExpr.bpl
test21/BooleanQuantification.bpl
test21/Boxing.bpl
test21/Casts.bpl
test21/Coercions2.bpl
test21/Colors.bpl
test21/DisjointDomains2.bpl
test21/EmptyList.bpl
test21/EmptySetBug.bpl
test21/FunAxioms.bpl
test21/HeapAxiom.bpl
test21/InterestingExamples0.bpl
test21/InterestingExamples2.bpl
test21/InterestingExamples3.bpl
test21/InterestingExamples4.bpl
test21/MapOutputTypeParams.bpl
test21/Maps0.bpl
test21/Maps1.bpl
test21/Orderings.bpl
test21/Orderings2.bpl
test21/Orderings3.bpl
test21/Orderings4.bpl
test21/PolyList.bpl
test21/Triggers0.bpl
test21/Triggers1.bpl
test7/MultipleErrors.bpl
```

**Diverging**
```
aitest0/Intervals.bpl
aitest9/TestIntervals.bpl
floats/git-issue80.bpl
houdini/houd5.bpl
houdini/houd6.bpl
houdini/mergedProgSingle_dac.bpl
inline/Elevator.bpl
inline/polyInline.bpl
inline/test5.bpl
livevars/bla1.bpl
livevars/daytona_bug2_ioctl_example_2.bpl
livevars/stack_overflow.bpl
prover/EQ_v2.Eval__v4.Eval_out.bpl
stratifiedinline/large.bpl
test2/Arrays.bpl
test2/Ensures.bpl
test2/Lambda.bpl
test2/LambdaLifting.bpl
test2/LambdaOldExpressions.bpl
test2/NullaryMaps.bpl
test2/Passification.bpl
test2/Timeouts0.bpl
test2/TypeEncodingM.bpl
test21/DisjointDomains.bpl
test21/EmptyList.bpl
test21/HeapAbstraction.bpl
test21/InterestingExamples1.bpl
test21/InterestingExamples2.bpl
test21/MapAxiomsConsistency.bpl
test21/MapOutputTypeParams.bpl
test21/Maps0.bpl
test21/ParallelAssignment.bpl
```

**Failing with current options**
```
extractloops/t3.bpl
test15/CaptureState.bpl
test15/ModelTest.bpl
test2/Rlimitouts0.bpl
test7/MultipleErrors.bpl
```

**Comment by bkragl** [comment]
I looked a bit into the code and it seems that there are the following ways to set prover options:

1) `/proverOpt` has three cases:
  1.1) Some generic prover-related options Boogie provides, like PROVER (to choose between Z3, CVC4, and YICES2), VERBOSITY, LOG_FILE, etc.
  1.2) Starting with `O:...` sets options for the SMT query.
  1.3) Starting with `C:...` sets a command-line argument for the prover process.
2) `/z3opt` takes options for Z3, which has the weird logic that @zvonimir referenced above to decide between query and command-line option.

I have the following proposal:

* Split the three cases of `/proverOpt`, for example:
  - `/proverOpt` for 1.1, `/proverPar` for 1.2, and `/proverArg` for 1.3.
  - `/proverOpt` for 1.2, `/proverArg` for 1.3, and separate "top-level" Boogie options for what's possible with 1.1.
* Eliminate `/z3opt`.

Since there is already a way for tools built on top of Boogie to pass their desired solver arguments, I think we can get rid of all the default options. For convenience, we can of course implement a feature to read default options from a file.

**Comment by zvonimir** [comment]
@bkragl could you please clarify what "diverging" means. Thx!

**Comment by zvonimir** [comment]
For what it's worth, both of your scary examples have polymorphic types. Not sure if that is the root cause of the mentioned weirdness.
Could you put your changes into a branch in this repo?
I can try to find some time to help with debugging.

**Comment by bkragl** [comment]
Diverging means that I killed the verifier after a couple of minutes (these tests are quite small and should verify quickly).

My changes are on the branch "z3-option-cleanup".

**Comment by shazqadeer** [comment]
@RustanLeino : Your input here would be helpful.  Do we expect the treatment of polymorphic features to be sound regardless of the options passed to Z3?  I also noticed that there is an option /typeEncoding:n which declares itself to be unsound.  Is it some experimental feature that should be tossed?  Should the entire handling of polymorphic features be tossed and re-implemented?

**Comment by zvonimir** [comment]
I've been playing with `bitvectors/bv5.bpl`. In particular, when I generate prover output using the master branch of Boogie (so with the currently used Z3 parameters), and then I run Z3 on the query, I get the following output:
```
unknown
(:reason-unknown "(incomplete quantifiers)")
(labels +3 +0 @2)
unsat
```
To me it seems that the reason why this benchmark is failing in the master branch is due to incompleteness of quantifier instantiation, and not the fact that Z3 found a model. I am not totally sure though.

**Comment by shazqadeer** [comment]
@bkragl, @zvonimir : What is your take on doing case analysis based on prover versions as is currently being done in the Boogie source code?

**Comment by zvonimir** [comment]
Sorry, but I am not sure I understand your proposal @shazqadeer. I just went through this discussion to reload it in my brain, and I am not sure that prover versions are critical for this. Maybe you could clarify your proposal further.

**Comment by shazqadeer** [comment]
I don't have a proposal.  I made an observation that another source of unpleasantness in Boogie's handling of SMT solvers is that it does specific things based on the version of the solver.  I was wondering whether this attribute of Boogie requires cleanup.  If yes, is this cleanup part of or orthogonal the discussion in this issue?

**Comment by zvonimir** [comment]
Ah, I see. Yeah, I would say that aspect should be cleaned up as well and it is related to this discussion (in my opinion).
The tricky part of this cleanup is that it will likely break some regressions, especially in terms of performance, and this almost certainly includes regressions of tools that are built on top of Boogie (e.g., Corral, Dafny). So the authors of these tools would have to implement some simple updates on their end.
I still think this should be done though.

**Comment by bkragl** [comment]
I agree. It would be best if we can reduce the default options to a bare minimum (which are hopefully among the stable ones in an evolving SMT solver) and eliminate any case analysis on the solver version. Except some recent changes in #189 (related to model_compress / model.compact), all other case distinctions are for Z3 versions that are anyway not supported anymore.
@michael-emmi: Was #189 done because Dafny regression testing uses Z3 older then 4.8.7? Then perhaps this could be solved by an upgrade. Or do you think we should keep this case analysis, at least for a while?

As I said [above](https://github.com/boogie-org/boogie/issues/73#issuecomment-565743197), in addition to reducing the set of default options, I think it would be good to clean up Boogie's command-line options related to provers. But then clients of Boogie have the ability to easily recover previous behavior in case something breaks for them. Indicating this breaking change by appropriately incrementing the version number and notifying all clients we know should be fine, right?

**Comment by michael-emmi** [comment]
> Was #189 done because Dafny regression testing uses Z3 older then 4.8.7? Then perhaps this could be solved by an upgrade.

@bkragl Yes. An alternative fix to #189 would be to upgrade Dafny’s Z3 version to 4.8.7. I introduced the version case split because the upgrade broke many more Dafny regressions than I had time to fix.
